libpromises/pipes_unix.c

```diff
@@ -142,6 +169,113 @@ static pid_t CreatePipeAndFork(const char *type, int *pd)
 
 /*****************************************************************************/
 
+static pid_t CreatePipeAndFork(const char *type, int *pd)
+{
+    IOPipe pipes[2] = {{0}};
+    pipes[0].type = type;
+    
+    pid_t pid = GenericCreatePipeAndFork(pipes);
+    
+    pd[0] = pipes[0].pipe_desc[0];
+    pd[1] = pipes[0].pipe_desc[1];
+
+    return pid;
+}
+
+/*****************************************************************************/
+
+static pid_t CreatePipesAndFork(const char *type, int *pd, int *pdb)
+{
+    IOPipe pipes[2] = {{0}};
+    /* Both pipes MUST have the same type. */
+    pipes[0].type = type;
+    pipes[1].type = type;
+
+    pid_t pid =  GenericCreatePipeAndFork(pipes);
+
+    pd[0] = pipes[0].pipe_desc[0];
+    pd[1] =  pipes[0].pipe_desc[1];
+    pdb[0] = pipes[1].pipe_desc[0];
+    pdb[1] = pipes[1].pipe_desc[1];
+    return pid;
+}
+
+/*****************************************************************************/
+
+IOData cf_popen_full_duplex(const char *command, bool capture_stderr)
+{
+/* For simplifying reading and writing directions */
+#define READ  0
+#define WRITE 1
+    int child_pipe[2];  /* From child to parent */
+    int parent_pipe[2]; /* From parent to child */
+    pid_t pid;
+
+    fflush(NULL); /* Empty file buffers */
+    pid = CreatePipesAndFork("rt", child_pipe, parent_pipe);
```

Comment:
Agree. In fact while opening bi-directional channel this is not used anywhere except the place where the check is performed if we have correct pipe type. I can leave it as "r" or "w" or can make the check to allow "r+t".